### DISCUSSION TITLE
How to handle ram in qemu?

### DISCUSSION THREAD

**Discussion by cekkr** [utmapp/UTM#4936] [Ideas]
Hello, 
I put here a question that I'm pretty sure [doesn't find an answer on stack overflow.](https://stackoverflow.com/questions/75102271/qemu-how-handling-ram-management)

I would like to create a personal algorithm for handling the ram of a virtual machine. There is the memory API: https://qemu.readthedocs.io/en/latest/devel/memory.html but I don't find any example for this particular use. Should I create a plugin? Or recompile qemu with it?

Then there is the possibility of directly changing memory management in QEMU. I think that this should be the file https://github.com/qemu/qemu/blob/266469947161aa10b1d36843580d369d5aa38589/hw/mem/memory-device.c.

I don't know what path to take to create an efficient handler and, at the same time, not mess up (too much) the code.

What do you think is the best solution?

Thank you

P.S. 
I would apply this algorithm right to UTM for Mac devices with just 8 gb of ram, maybe if it works well I can do a pull request

**Comment by saagarjha** [comment]
What problem are you trying to solve?

**Reply by cekkr** [reply]
Practically qemu seems to ignore when the guest OS free some memory area. So I would like to create a sort of paging system at the host level that gives priority to the most accessed regions. 

But eventually, I would like to learn its functioning in any case.

**Reply by saagarjha** [reply]
Might be relevant? https://github.com/qemu/qemu/blob/master/hw/virtio/virtio-balloon.c

**Reply by cekkr** [reply]
> https://github.com/qemu/qemu/blob/master/hw/virtio/virtio-balloon.c

But in this way the guest doesn't lose the released region?


In any case, just now I thought the we are not talking about emulation, but virtualization on Apple Silicon: is possible in virtualization to handle with a middleware the ram?
